File: src/analysis/offline_spectrum.c
#include "sounds/offline_spectrum.h"

#include <Accelerate/Accelerate.h>

#include <math.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

static const double pi_value = 3.14159265358979323846264338327950288;
static const double log_two = 0.693147180559945309417232121458176568;
static const double display_db_floor = -140.0;
static const double minimum_amplitude = 1.0e-7;

typedef struct RealDft {
    uint64_t length;
    uint64_t half_length;
    float *time;
    float *even;
    float *odd;
    float *real;
    float *imag;
    vDSP_DFT_Setup setup;
} RealDft;

static double clamp_double(double value, double minimum, double maximum) {
    if (value < minimum) {
        return minimum;
    }

    if (value > maximum) {
        return maximum;
    }

    return value;
}

static double log2_double(double value) {
    return log(value) / log_two;
}

static bool multiply_overflows_size(uint64_t count, size_t element_size) {
    return count > (uint64_t)(SIZE_MAX / element_size);
}

static bool allocate_float_buffer(float **buffer, uint64_t count) {
    if (multiply_overflows_size(count, sizeof(float))) {
        return false;
    }

    *buffer = calloc((size_t)count, sizeof(float));
    return *buffer != NULL;
}

static bool next_power_of_two(uint64_t value, uint64_t *result) {
    uint64_t power = 2;

    while (power < value) {
        if (power > UINT64_MAX / 2U) {
            return false;
        }

        power *= 2U;
    }

    *result = power;
    return true;
}

static void real_dft_destroy(RealDft *dft) {
    if (!dft) {
        return;
    }

    vDSP_DFT_DestroySetup(dft->setup);
    free(dft->time);
    free(dft->even);
    free(dft->odd);
    free(dft->real);
    free(dft->imag);
    memset(dft, 0, sizeof(*dft));
}

static bool real_dft_create(RealDft *dft, uint64_t length, SoundError *error) {
    memset(dft, 0, sizeof(*dft));

    if (length < 2U || (length & 1U) != 0U) {
        sound_error_set(error, "invalid offline spectrum DFT length");
        return false;
    }

    dft->length = length;
    dft->half_length = length / 2U;
    dft->setup = vDSP_DFT_zrop_CreateSetup(NULL, (vDSP_Length)length, vDSP_DFT_FORWARD);

    if (!dft->setup ||
        !allocate_float_buffer(&dft->time, length) ||
        !allocate_float_buffer(&dft->even, dft->half_length) ||
        !allocate_float_buffer(&dft->odd, dft->half_length) ||
        !allocate_float_buffer(&dft->real, dft->half_length) ||
        !allocate_float_buffer(&dft->imag, dft->half_length)) {
        real_dft_destroy(dft);
        sound_error_set(error, "could not allocate offline spectrum DFT buffers");
        return false;
    }

    return true;
}

static void real_dft_forward(RealDft *dft) {
    for (uint64_t i = 0; i < dft->half_length; ++i) {
        dft->even[i] = dft->time[i * 2U];
        dft->odd[i] = dft->time[i * 2U + 1U];
    }

    vDSP_DFT_Execute(dft->setup, dft->even, dft->odd, dft->real, dft->imag);
}

static double row_edge_frequency(
    double min_hz,
    double max_hz,
    uint64_t edge,
    uint64_t row_count
) {
    double unit = row_count == 0U ? 0.0 : (double)edge / (double)row_count;
    double log_min = log2_double(min_hz);
    double log_max = log2_double(max_hz);
    double log_hz = log_max + (log_min - log_max) * clamp_double(unit, 0.0, 1.0);

    return pow(2.0, log_hz);
}

static double bin_power(const RealDft *dft, uint64_t bin) {
    double real;
    double imag;

    if (bin == 0U) {
        real = dft->real[0];
        imag = 0.0;
    } else if (bin >= dft->half_length) {
        real = dft->imag[0];
        imag = 0.0;
    } else {
        real = dft->real[bin];
        imag = dft->imag[bin];
    }

    return real * real + imag * imag;
}

static float power_to_db(double power, double window_sum, uint64_t bin, uint64_t half_length) {
    double amplitude = sqrt(fmax(power, 0.0)) / fmax(window_sum, 1.0e-12);

    if (bin > 0U && bin < half_length) {
        amplitude *= 2.0;
    }

    return (float)fmax(display_db_floor, 20.0 * log10(fmax(amplitude, minimum_amplitude)));
}

double sound_offline_spectrum_frequency_for_row(
    double min_hz,
    double max_hz,
    uint64_t row,
    uint64_t row_count
) {
    if (min_hz <= 0.0 || max_hz <= min_hz || row_count == 0U) {
        return 0.0;
    }

    double unit = row_count == 1U ? 0.5 : ((double)row + 0.5) / (double)row_count;
    double log_min = log2_double(min_hz);
    double log_max = log2_double(max_hz);
    double log_hz = log_max + (log_min - log_max) * clamp_double(unit, 0.0, 1.0);

    return pow(2.0, log_hz);
}

bool sound_offline_spectrum_db(
    const float *samples,
    uint64_t sample_count,
    double sample_rate,
    double min_hz,
    double max_hz,
    float *dbfs_rows,
    uint64_t row_count,
    SoundError *error
) {
    sound_error_clear(error);

    if (!samples || !dbfs_rows || sample_count == 0U || row_count == 0U ||
        sample_rate <= 0.0 || min_hz <= 0.0 || max_hz <= min_hz) {
        sound_error_set(error, "invalid offline spectrum request");
        return false;
    }

    double nyquist = sample_rate * 0.5;
    max_hz = fmin(max_hz, nyquist);
    if (min_hz >= max_hz) {
        sound_error_set(error, "offline spectrum range is outside Nyquist");
        return false;
    }

    uint64_t fft_length = 0;
    if (!next_power_of_two(sample_count, &fft_length)) {
        sound_error_set(error, "offline spectrum clip is too large");
        return false;
    }

    RealDft dft;
    if (!real_dft_create(&dft, fft_length, error)) {
        return false;
    }

    double window_sum = 0.0;
    if (sample_count == 1U) {
        dft.time[0] = samples[0];
        window_sum = 1.0;
    } else {
        for (uint64_t i = 0; i < sample_count; ++i) {
            double unit = (double)i / (double)(sample_count - 1U);
            double window = 0.5 - 0.5 * cos(2.0 * pi_value * unit);

            dft.time[i] = (float)((double)samples[i] * window);
            window_sum += window;
        }
    }

    real_dft_forward(&dft);

    for (uint64_t row = 0; row < row_count; ++row) {
        double high_hz = row_edge_frequency(min_hz, max_hz, row, row_count);
        double low_hz = row_edge_frequency(min_hz, max_hz, row + 1U, row_count);
        uint64_t first_bin = (uint64_t)floor(low_hz * (double)fft_length / sample_rate);
        uint64_t last_bin = (uint64_t)ceil(high_hz * (double)fft_length / sample_rate);

        if (first_bin > dft.half_length) {
            first_bin = dft.half_length;
        }

        if (last_bin > dft.half_length) {
            last_bin = dft.half_length;
        }

        if (last_bin < first_bin) {
            last_bin = first_bin;
        }

        double best_power = 0.0;
        uint64_t best_bin = first_bin;
        for (uint64_t bin = first_bin; bin <= last_bin; ++bin) {
            double power = bin_power(&dft, bin);
            if (power > best_power) {
                best_power = power;
                best_bin = bin;
            }

            if (bin == dft.half_length) {
                break;
            }
        }

        dbfs_rows[row] = power_to_db(best_power, window_sum, best_bin, dft.half_length);
    }

    real_dft_destroy(&dft);
    return true;
}

File: src/analysis/offline_spectrum.zig
//! Whole-clip spectrum and spectrogram analysis.

const std = @import("std");
const vdsp = @import("../apple/vdsp.zig");

pub const Error = error{
    InvalidRequest,
    InvalidRange,
    SetupFailed,
    TooLarge,
};

const pi_value = std.math.pi;
const display_db_floor = -140.0;
const minimum_amplitude = 1.0e-7;
const maximum_offline_sample_rate = 512000.0;
const spectrogram_min_window = 1024;
const spectrogram_max_window = 16384;

const RealDft = struct {
    allocator: std.mem.Allocator,
    length: usize,
    half_length: usize,
    time: []f32,
    even: []f32,
    odd: []f32,
    real: []f32,
    imag: []f32,
    setup: vdsp.DFTSetup,

    fn init(allocator: std.mem.Allocator, length: usize) !RealDft {
        if (length < 2 or (length & 1) != 0) return Error.InvalidRequest;

        const half = length / 2;
        var dft = RealDft{
            .allocator = allocator,
            .length = length,
            .half_length = half,
            .time = &.{},
            .even = &.{},
            .odd = &.{},
            .real = &.{},
            .imag = &.{},
            .setup = null,
        };
        errdefer dft.deinit();

        dft.setup = vdsp.createDft(null, length, .forward) catch return Error.SetupFailed;
        dft.time = try allocator.alloc(f32, length);
        dft.even = try allocator.alloc(f32, half);
        dft.odd = try allocator.alloc(f32, half);
        dft.real = try allocator.alloc(f32, half);
        dft.imag = try allocator.alloc(f32, half);
        @memset(dft.time, 0.0);

        return dft;
    }

    fn deinit(self: *RealDft) void {
        vdsp.destroyDft(self.setup);
        self.allocator.free(self.time);
        self.allocator.free(self.even);
        self.allocator.free(self.odd);
        self.allocator.free(self.real);
        self.allocator.free(self.imag);
        self.* = undefined;
    }

    fn forward(self: *RealDft) void {
        const lanes = comptime (std.simd.suggestVectorLength(f32) orelse 4);
        const Vec = @Vector(lanes, f32);
        var index: usize = 0;
        while (index + lanes <= self.half_length) : (index += lanes) {
            var even: Vec = undefined;
            var odd: Vec = undefined;
            inline for (0..lanes) |lane| {
                even[lane] = self.time[(index + lane) * 2];
                odd[lane] = self.time[(index + lane) * 2 + 1];
            }
            self.even[index..][0..lanes].* = even;
            self.odd[index..][0..lanes].* = odd;
        }
        while (index < self.half_length) : (index += 1) {
            self.even[index] = self.time[index * 2];
            self.odd[index] = self.time[index * 2 + 1];
        }

        vdsp.executeDft(self.setup, self.even, self.odd, self.real, self.imag);
    }

    fn binPower(self: *const RealDft, bin: usize) f64 {
        if (bin == 0) {
            return @as(f64, self.real[0]) * @as(f64, self.real[0]);
        }

        if (bin >= self.half_length) {
            return @as(f64, self.imag[0]) * @as(f64, self.imag[0]);
        }

        return @as(f64, self.real[bin]) * @as(f64, self.real[bin]) +
            @as(f64, self.imag[bin]) * @as(f64, self.imag[bin]);
    }
};

pub fn frequencyForRow(min_hz: f64, max_hz: f64, row: usize, row_count: usize) f64 {
    if (min_hz <= 0.0 or max_hz <= min_hz or row_count == 0) return 0.0;

    const unit = if (row_count == 1)
        0.5
    else
        (@as(f64, @floatFromInt(row)) + 0.5) / @as(f64, @floatFromInt(row_count));
    const log_min = @log2(min_hz);
    const log_max = @log2(max_hz);
    const log_hz = log_max + (log_min - log_max) * std.math.clamp(unit, 0.0, 1.0);

    return std.math.pow(f64, 2.0, log_hz);
}

pub fn spectrumDb(
    allocator: std.mem.Allocator,
    samples: []const f32,
    sample_rate: f64,
    min_hz: f64,
    max_hz_arg: f64,
    dbfs_rows: []f32,
) !void {
    if (samples.len == 0 or
        dbfs_rows.len == 0 or
        !std.math.isFinite(sample_rate) or
        sample_rate <= 0.0 or
        sample_rate > maximum_offline_sample_rate or
        min_hz <= 0.0 or
        max_hz_arg <= min_hz)
    {
        return Error.InvalidRequest;
    }

    const nyquist = sample_rate * 0.5;
    const max_hz = @min(max_hz_arg, nyquist);
    if (min_hz >= max_hz) return Error.InvalidRange;

    const fft_length = try nextPowerOfTwo(samples.len);
    var dft = try RealDft.init(allocator, fft_length);
    defer dft.deinit();

    @memset(dft.time, 0.0);
    var window_sum: f64 = 0.0;
    if (samples.len == 1) {
        dft.time[0] = samples[0];
        window_sum = 1.0;
    } else {
        for (samples, 0..) |sample, i| {
            const unit = @as(f64, @floatFromInt(i)) /
                @as(f64, @floatFromInt(samples.len - 1));
            const window = 0.5 - 0.5 * @cos(2.0 * pi_value * unit);

            dft.time[i] = @floatCast(@as(f64, sample) * window);
            window_sum += window;
        }
    }

    dft.forward();
    rowsFromDft(&dft, sample_rate, min_hz, max_hz, window_sum, dbfs_rows);
}

pub fn spectrogramDb(
    allocator: std.mem.Allocator,
    samples: []const f32,
    sample_rate: f64,
    min_hz: f64,
    max_hz_arg: f64,
    dbfs_columns: []f32,
    row_count: usize,
    column_count: usize,
) !void {
    if (samples.len == 0 or
        row_count == 0 or
        column_count == 0 or
        dbfs_columns.len < row_count * column_count or
        !std.math.isFinite(sample_rate) or
        sample_rate <= 0.0 or
        sample_rate > maximum_offline_sample_rate or
        min_hz <= 0.0 or
        max_hz_arg <= min_hz)
    {
        return Error.InvalidRequest;
    }

    if (column_count > std.math.maxInt(usize) / row_count) return Error.TooLarge;

    const nyquist = sample_rate * 0.5;
    const max_hz = @min(max_hz_arg, nyquist);
    if (min_hz >= max_hz) return Error.InvalidRange;

    const fft_length = try spectrogramWindowLength(samples.len, column_count);
    var dft = try RealDft.init(allocator, fft_length);
    defer dft.deinit();

    for (0..column_count) |column| {
        const center = spectrogramColumnCenter(samples.len, column, column_count);
        const window_sum = fillCenteredWindow(&dft, samples, center);
        dft.forward();
        rowsFromDft(
            &dft,
            sample_rate,
            min_hz,
            max_hz,
            window_sum,
            dbfs_columns[column * row_count ..][0..row_count],
        );
    }
}

pub fn bandAmplitudeOverTime(
    allocator: std.mem.Allocator,
    samples: []const f32,
    sample_rate: f64,
    low_hz: f64,
    high_hz_arg: f64,
    amplitudes: []f32,
) !void {
    if (samples.len == 0 or
        amplitudes.len == 0 or
        !std.math.isFinite(sample_rate) or
        sample_rate <= 0.0 or
        sample_rate > maximum_offline_sample_rate or
        !std.math.isFinite(low_hz) or
        !std.math.isFinite(high_hz_arg) or
        low_hz <= 0.0 or
        high_hz_arg <= low_hz)
    {
        return Error.InvalidRequest;
    }

    const nyquist = sample_rate * 0.5;
    const high_hz = @min(high_hz_arg, nyquist);
    if (low_hz >= high_hz) return Error.InvalidRange;

    const fft_length = try spectrogramWindowLength(samples.len, amplitudes.len);
    var dft = try RealDft.init(allocator, fft_length);
    defer dft.deinit();

    const bin_range = bandBinRange(fft_length, dft.half_length, sample_rate, low_hz, high_hz);
    for (amplitudes, 0..) |*amplitude, column| {
        const center = spectrogramColumnCenter(samples.len, column, amplitudes.len);
        const window_sum = fillCenteredWindow(&dft, samples, center);
        dft.forward();
        amplitude.* = integratedBandAmplitude(&dft, bin_range.first, bin_range.last, window_sum);
    }
}

fn rowsFromDft(
    dft: *const RealDft,
    sample_rate: f64,
    min_hz: f64,
    max_hz: f64,
    window_sum: f64,
    dbfs_rows: []f32,
) void {
    for (dbfs_rows, 0..) |*db, row| {
        const high_hz = rowEdgeFrequency(min_hz, max_hz, row, dbfs_rows.len);
        const low_hz = rowEdgeFrequency(min_hz, max_hz, row + 1, dbfs_rows.len);
        var first_bin: usize = @intFromFloat(@floor(low_hz * @as(f64, @floatFromInt(dft.length)) / sample_rate));
        var last_bin: usize = @intFromFloat(@ceil(high_hz * @as(f64, @floatFromInt(dft.length)) / sample_rate));

        if (first_bin > dft.half_length) first_bin = dft.half_length;
        if (last_bin > dft.half_length) last_bin = dft.half_length;
        if (last_bin < first_bin) last_bin = first_bin;

        var best_power: f64 = 0.0;
        var best_bin = first_bin;
        var bin = first_bin;
        while (bin <= last_bin) : (bin += 1) {
            const power = dft.binPower(bin);
            if (power > best_power) {
                best_power = power;
                best_bin = bin;
            }

            if (bin == dft.half_length) break;
        }

        db.* = powerToDb(best_power, window_sum, best_bin, dft.half_length);
    }
}

fn bandBinRange(
    fft_length: usize,
    half_length: usize,
    sample_rate: f64,
    low_hz: f64,
    high_hz: f64,
) struct { first: usize, last: usize } {
    const length_f = @as(f64, @floatFromInt(fft_length));
    var first: usize = @intFromFloat(@floor(low_hz * length_f / sample_rate));
    var last: usize = @intFromFloat(@ceil(high_hz * length_f / sample_rate));

    if (first > half_length) first = half_length;
    if (last > half_length) last = half_length;
    if (last < first) last = first;
    return .{ .first = first, .last = last };
}

fn integratedBandAmplitude(
    dft: *const RealDft,
    first_bin: usize,
    last_bin: usize,
    window_sum: f64,
) f32 {
    var power: f64 = 0.0;
    var bin = first_bin;
    while (bin <= last_bin) : (bin += 1) {
        const scale: f64 = if (bin > 0 and bin < dft.half_length) 4.0 else 1.0;
        power += dft.binPower(bin) * scale;
        if (bin == dft.half_length) break;
    }

    return @floatCast(@sqrt(@max(power, 0.0)) / @max(window_sum, 1.0e-12));
}

fn powerToDb(power: f64, window_sum: f64, bin: usize, half_length: usize) f32 {
    var amplitude = @sqrt(@max(power, 0.0)) / @max(window_sum, 1.0e-12);

    if (bin > 0 and bin < half_length) amplitude *= 2.0;

    return @floatCast(@max(
        display_db_floor,
        20.0 * @log10(@max(amplitude, minimum_amplitude)),
    ));
}

fn rowEdgeFrequency(min_hz: f64, max_hz: f64, edge: usize, row_count: usize) f64 {
    const unit = if (row_count == 0)
        0.0
    else
        @as(f64, @floatFromInt(edge)) / @as(f64, @floatFromInt(row_count));
    const log_min = @log2(min_hz);
    const log_max = @log2(max_hz);
    const log_hz = log_max + (log_min - log_max) * std.math.clamp(unit, 0.0, 1.0);

    return std.math.pow(f64, 2.0, log_hz);
}

fn spectrogramWindowLength(sample_count: usize, column_count: usize) !usize {
    if (sample_count > std.math.maxInt(usize) - column_count + 1) return Error.TooLarge;

    const samples_per_column = (sample_count + column_count - 1) / column_count;
    if (samples_per_column > std.math.maxInt(usize) / 4) return Error.TooLarge;

    const wanted = std.math.clamp(
        samples_per_column * 4,
        spectrogram_min_window,
        spectrogram_max_window,
    );
    var length = try nextPowerOfTwo(wanted);

    if (length > spectrogram_max_window) length = spectrogram_max_window;
    return length;
}

fn spectrogramColumnCenter(sample_count: usize, column: usize, column_count: usize) usize {
    const numerator =
        (@as(f128, @floatFromInt(column)) * 2.0 + 1.0) *
        @as(f128, @floatFromInt(sample_count));
    const denominator = 2.0 * @as(f128, @floatFromInt(column_count));
    const center = numerator / denominator;

    if (center >= @as(f128, @floatFromInt(sample_count))) {
        return if (sample_count > 0) sample_count - 1 else 0;
    }

    return @intFromFloat(center);
}

fn fillCenteredWindow(dft: *RealDft, samples: []const f32, center: usize) f64 {
    @memset(dft.time, 0.0);

    const half = dft.length / 2;
    var window_sum: f64 = 0.0;

    for (0..dft.length) |i| {
        const unit = if (dft.length == 1)
            0.0
        else
            @as(f64, @floatFromInt(i)) / @as(f64, @floatFromInt(dft.length - 1));
        const window = 0.5 - 0.5 * @cos(2.0 * pi_value * unit);
        var sample_index: usize = 0;
        var in_range = false;

        if (i >= half) {
            const offset = i - half;
            if (center <= std.math.maxInt(usize) - offset) {
                sample_index = center + offset;
                in_range = sample_index < samples.len;
            }
        } else {
            const offset = half - i;
            if (center >= offset) {
                sample_index = center - offset;
                in_range = true;
            }
        }

        if (in_range) {
            dft.time[i] = @floatCast(@as(f64, samples[sample_index]) * window);
        }

        window_sum += window;
    }

    return window_sum;
}

fn nextPowerOfTwo(value: usize) !usize {
    var power: usize = 2;

    while (power < value) {
        if (power > std.math.maxInt(usize) / 2) return Error.TooLarge;
        power *= 2;
    }

    return power;
}

test "offline spectrum maps rows and produces finite tone energy" {
    const sample_rate = 48000.0;
    const sample_count = 4096;
    var input: [sample_count]f32 = undefined;
    var rows: [96]f32 = undefined;

    for (&input, 0..) |*sample, i| {
        const t = @as(f64, @floatFromInt(i)) / sample_rate;
        sample.* = @floatCast(@sin(2.0 * pi_value * 500.0 * t));
    }

    try spectrumDb(std.testing.allocator, &input, sample_rate, 20.0, 24000.0, &rows);

    var maximum: f32 = -1.0e30;
    for (rows) |row| {
        try std.testing.expect(std.math.isFinite(row));
        maximum = @max(maximum, row);
    }

    try std.testing.expect(maximum > -20.0);
    try std.testing.expect(frequencyForRow(20.0, 24000.0, 0, 96) > frequencyForRow(20.0, 24000.0, 95, 96));
}

test "band amplitude envelope follows in-band burst energy" {
    const sample_rate = 48000.0;
    const seconds = 2.0;
    const sample_count: usize = @intFromFloat(sample_rate * seconds);
    const bucket_count = 96;

    const input = try std.testing.allocator.alloc(f32, sample_count);
    defer std.testing.allocator.free(input);
    var low_band: [bucket_count]f32 = undefined;
    var high_band: [bucket_count]f32 = undefined;

    for (input, 0..) |*sample, i| {
        const t = @as(f64, @floatFromInt(i)) / sample_rate;
        var value = 0.35 * @sin(2.0 * pi_value * 500.0 * t);
        if (i >= sample_count / 3 and i < (sample_count * 2) / 3) {
            value += 0.45 * @sin(2.0 * pi_value * 4000.0 * t);
        }
        sample.* = @floatCast(value);
    }

    try bandAmplitudeOverTime(std.testing.allocator, input, sample_rate, 3800.0, 4200.0, &high_band);
    try bandAmplitudeOverTime(std.testing.allocator, input, sample_rate, 450.0, 550.0, &low_band);

    const Helpers = struct {
        fn mean(values: []const f32, first: usize, last: usize) f64 {
            var sum: f64 = 0.0;
            for (values[first..last]) |value| sum += value;
            return sum / @as(f64, @floatFromInt(last - first));
        }
    };

    const first_high = Helpers.mean(&high_band, 0, bucket_count / 3);
    const middle_high = Helpers.mean(&high_band, bucket_count / 3, (bucket_count * 2) / 3);
    const last_high = Helpers.mean(&high_band, (bucket_count * 2) / 3, bucket_count);
    try std.testing.expect(middle_high > first_high * 5.0);
    try std.testing.expect(middle_high > last_high * 5.0);

    const first_low = Helpers.mean(&low_band, 4, bucket_count / 3);
    const middle_low = Helpers.mean(&low_band, bucket_count / 3, (bucket_count * 2) / 3);
    const last_low = Helpers.mean(&low_band, (bucket_count * 2) / 3, bucket_count - 4);
    try std.testing.expectApproxEqRel(first_low, middle_low, 0.25);
    try std.testing.expectApproxEqRel(last_low, middle_low, 0.25);
}
